Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+```swift
+/// The LambdaRuntime object. This object communicates with the Lambda control plane
+/// to fetch work and report errors.
+public final class LambdaRuntime<Handler>: ServiceLifecycle.Service, Sendable
+    where Handler: StreamingLambdaHandler, Handler: ~Copyable
+{
+
+    /// Create a LambdaRuntime by passing a handler, an eventLoop and a logger.
+    /// - Parameter handler: A ``StreamingLambdaHandler`` that will be invoked
+    /// - Parameter eventLoop: An ``EventLoop`` on which the LambdaRuntime will be
+    ///                        executed. Defaults to an EventLoop from
+    ///                        ``NIOSingletons.posixEventLoopGroup``.
+    /// - Parameter logger: A logger
+    public init(
+        handler: consuming sending Handler,
+        eventLoop: EventLoop = Lambda.defaultEventLoop,
+        logger: Logger = Logger(label: "Lambda")
+    )
+
+    /// Create a LambdaRuntime by passing a ``StreamingLambdaHandler``.
+    public convenience init(handler: consuming sending Handler)
+
+    /// Starts the LambdaRuntime by connecting to the Lambda control plane to ask
+    /// for events to process. If the environment variable AWS_LAMBDA_RUNTIME_API is
+    /// set, the LambdaRuntime will connect to the Lambda control plane. Otherwise
+    /// it will start a mock server that can be used for testing at port 8080
+    /// locally.
+    /// Cancel the task that runs this function to close the communication with
+    /// the Lambda control plane or close the local mock server. This function
+    /// only returns once cancelled.
+    public func run() async throws
+}
+```
+
+The current API allows for a Lambda function to be tested locally through a mock server by requiring an environment variable named  `LOCAL_LAMBDA_SERVER_ENABLED` to be set to `true`. If this environment variable is not set, the program immediately crashes as the user will not have the `AWS_LAMBDA_RUNTIME_API` environment variable on their local machine (set automatically when deployed to AWS Lambda). However, making the user set the `LOCAL_LAMBDA_SERVER_ENABLED` environment variable is an unnecessary step that can be avoided. In the v2 API, the `run()` function will automatically start the mock server when the `AWS_LAMBDA_RUNTIME_API` environment variable cannot be found.
+
+### Lambda
+
+We also add an enum to store a static function and a property on. We put this on the static `Lambda` because `LambdaRuntime` is generic and thus has bad ergonomics for static properties and functions.
+
+```swift
+enum Lambda {
+    /// This returns the default EventLoop that a LambdaRuntime is scheduled on.
+    /// It uses `NIOSingletons.posixEventLoopGroup.next()` under the hood.
+    public static var defaultEventLoop: any EventLoop { get }
+
+    /// Report a startup error to the Lambda Control Plane API
+    public static func reportStartupError(any Error) async
```

Comment:
What if we make this a `with` method so that users can do any setup and we report any error that happens there. Instead of them having to do a `do/catch`?